Support for pruned scan for required fields.

HyukjinKwon opened the pull request:
https://github.com/databricks/spark-csv/issues/166

Firstly,
I tested the pure iteration time of produced RDD from `buildScan()` with `lineitem` table in TPC-H.
On local machine, I set the number of 1,000,000 rows and measured the time.

The result was below. (The total number of columns was 17)

|Scan Type|Required Fields|Elapsed Time|
|--------|-------|-------|
|`PrunedScan`|`C0`|0.106|
|`PrunedScan`|`C0, C1, C2, C3, C4, C5, C6`|0.287|
|`PrunedScan`|`* (C0 - C16)`|0.505|
|`TableScan`|`C0`|0.536|
|`TableScan`|`C0, C1, C2, C3, C4, C5, C6`|0.502|
|`TableScan`|`* (C0 - C16)`|0.489|

Secondly, I didn't add test codes separately since the original test codes cover this.

Could you review this please?

JoshRosen commented:
Cool! I'll take a closer look at this tomorrow.

JoshRosen commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,58 @@ case class CsvRelation protected[spark] (
+          } else {
+            tokens
+          }
+          var index: Int = 0
```
Minor code simplification: I think that `index` and `subIndex` can be declared inside of the `try` block in order to use two fewer lines of code.

HyukjinKwon replied:
oh right thanks!

JoshRosen commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,58 @@ case class CsvRelation protected[spark] (
+          }
+          var index: Int = 0
+          var subIndex: Int = 0
+          val rowArray = new Array[Any](requiredIndices.length)
```
If you're interested in leveraging mutability as a performance optimization, then I think you might be able to re-use a single mutable array rather than creating a new one for each row.

JoshRosen replied:
On the other hand, the non-pruned-scan codepath didn't do this, so we should leave this optimization to a followup.

HyukjinKwon replied:
I though the use of a mutable array decreases performance.
Thanks, I will update them soon.

JoshRosen commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,58 @@ case class CsvRelation protected[spark] (
+        } else if (failFast && schemaFields.length != tokens.size) {
+          throw new RuntimeException(s"Malformed line in FAILFAST mode: ${tokens.mkString(",")}")
+        } else {
+          val indexSafeTokens = if (permissive && schemaFields.length != tokens.size) {
```
If `tokens` contains fewer columns than expected and we're _not_ in permissive mode, then should we throw a more explicit error rather than allowing an ArrayIndexOutOfBounds to occur?

JoshRosen replied:
Oh, nevermind, I see that this is covered by the `if (failFast && schemaFields.length != tokens.size)` case above.

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,56 @@ case class CsvRelation protected[spark] (
+      buildScan
+    } else {
+      val requiredIndices = new Array[Int](requiredFields.length)
+      schemaFields.zipWithIndex.filter(pair => requiredFields.contains(pair._1))
```
Nit: would be more readable if you match the tuple:
```scala
schemaFields.zipWithIndex.filter { 
  case (field, _) requiredFields.contains(field)
}
```

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,56 @@ case class CsvRelation protected[spark] (
+    } else {
+      val requiredIndices = new Array[Int](requiredFields.length)
+      schemaFields.zipWithIndex.filter(pair => requiredFields.contains(pair._1))
+        .foreach(pair => requiredIndices.update(requiredFields.indexOf(pair._1), pair._2))
```
Same here. For better readability please match the tuple inside foreach: ```foreach { case(field, index) => ... }```

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,56 @@ case class CsvRelation protected[spark] (
+      val rowArray = new Array[Any](requiredIndices.length)
+      tokenRdd(schemaFields.map(_.name)).flatMap { tokens =>
+        if (dropMalformed && schemaFields.length != tokens.size) {
+          logger.warn(s"Dropping malformed line: ${tokens.mkString(",")}")
```
Would be better to log the actual delimiter:
```scala
logger.warn(s"Dropping malformed line: ${tokens.mkString(delimiter)}")
```

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,56 @@ case class CsvRelation protected[spark] (
+          logger.warn(s"Dropping malformed line: ${tokens.mkString(",")}")
+          None
+        } else if (failFast && schemaFields.length != tokens.size) {
+          throw new RuntimeException(s"Malformed line in FAILFAST mode: ${tokens.mkString(",")}")
```
Same here.

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,56 @@ case class CsvRelation protected[spark] (
+            while (subIndex < requiredIndices.length) {
+              index = requiredIndices(subIndex)
+              val field = schemaFields(index)
+              rowArray(subIndex) = TypeCast.castTo(indexSafeTokens(index), field.dataType,
```
style nit:
```scala
rowArray(subIndex) = TypeCast.castTo(
    indexSafeTokens(index), 
    field.dataType,
    field.nullable,
    treatEmptyValuesAsNulls)
```

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,56 @@ case class CsvRelation protected[spark] (
+          } catch {
+            case nfe: java.lang.NumberFormatException if dropMalformed =>
+              logger.warn("Number format exception. " +
+                s"Dropping malformed line: ${tokens.mkString(",")}")
```
Same here with original delimiter.

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,56 @@ case class CsvRelation protected[spark] (
+              None
+            case pe: java.text.ParseException =>
+              logger.warn("Parse Exception. " +
+                s"Dropping malformed line: ${tokens.mkString(",")}")
```
and here.

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -136,6 +136,56 @@ case class CsvRelation protected[spark] (
     }
   }
 
+  override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
```
Would you please add some comments to this method and explain the general logic? This will help other developers quickly understand what is going on when they want to contribute to this code.

HyukjinKwon replied:
Right! I will add comments like the ones in `org.apache.spark.sql.sources.PrunedScan` at Spark
Also, let me update them (below) soon.

falaki commented:
@HyukjinKwon Thanks for adding this feature. I left some comments, mostly regarding style and readability.

HyukjinKwon commented:
@falaki Thanks for your advices. I updated them.
Also, I changed `requiredIndices.update(requiredFields.indexOf(field), index)` to `requiredIndices(requiredFields.indexOf(field)) = index`,
and `java.text.ParseException` to `java.text.ParseException if dropMalformed =>` which I think I removed  by mistake while testing.
